File: src/Page.java
package src;

import java.io.File;
import java.io.IOException;
import java.io.FileWriter;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Page {
    private final Settings settings;
    private File file;
    private String[] body;
    private String title;
    private String id;
    private LocalDate createdDate;  // this is actual the first appearance in the logs, not the date the page was created.
    private LocalDate updatedDate;
    private ArrayList<LogEntry> logs = new ArrayList<LogEntry>();
    private String section;
    final String NL = System.getProperty("line.separator");

    public Page(File f, Settings s) {
        this.file = f;
        this.settings = s;
        //  need to put file contents into body
        if (file != null) {
            try(Scanner sc = new Scanner(file)) {
                List<String> lines = new ArrayList<String>();
                while (sc.hasNextLine()) {
                    lines.add(sc.nextLine());
                }
                body = lines.toArray(new String[0]);
                // String[] arr = lines.toArray(new String[0]);  // why new String[0]? https://docs.oracle.com/javase/8/docs/api/java/util/List.html#toArray-T:A-
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        setIDFromFilename();
        setSectionFromFilename();
        this.createdDate = LocalDate.now();
        this.updatedDate = LocalDate.now();
        parsePiHeader(body[0]);
    }

    public Page(String[] body, Settings s) {
        //` FIXME this is hacky to get index page working
        this.settings = s;
        this.body = body;
        this.createdDate = LocalDate.now();
        this.updatedDate = LocalDate.now();
        parsePiHeader(body[0]);
    }

    private void setIDFromFilename() {
        // returns filename without extension
        String fileName = file.getName();
        int pos = fileName.lastIndexOf(".");
        if (pos > 0 && pos < (fileName.length() - 1)) { // If '.' is not the first or last character.
            fileName = fileName.substring(0, pos);
        }
        id = fileName;
    }

    private void setSectionFromFilename() {
        section = file.getParentFile().getName();
    }

    public void setBody(String[] body) {
        this.body = body;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public void setID(String id) {
        this.id = id;
    }

    public String getID() {
        return id;
    }

    public void setSection(String section) {
        this.section = section;
    }

    public String getSection() {
        return section;
    }

    public void addLogEntry(LogEntry l) {
        logs.add(l);
        // System.out.println("Adding log entry: " + l.toString());
    }

    private void parsePiHeader(String header) {
        /* Should be { key: value, key: value, key: value } */
        header = header.replaceAll("\\{", "").replaceAll("\\}", "");
        String[] options = header.split(",");
        for (String option : options) {
            String[] kv = option.split(":");
            switch (kv[0].trim()) {
                case "title":
                    setTitle(kv[1].trim());
                    break;
                default:
                    System.out.println("Header key " + kv[0].trim() + " refers to non-existent page property.");
                    break;
            }
        }
    }

    private String parsePiContent() {
        int lineNum = 0;
        String s = "";    

        lineNum = settings.numHeaderLines;

        ParseMode mode = ParseMode.NORMAL;
            
        for(; lineNum < body.length; ++lineNum) {
            String line = body[lineNum];
            // parse bold/italics
            line = parsePiTags(line, "\\*\\*(.*?)\\*\\*", "strong");
            line = parsePiTags(line, "__(.*?)__", "em");
            line = parsePiTags(line, "~~(.*?)~~", "strike");

            // parse URLs
            line = parsePiLinks(line);

            if ((mode == ParseMode.PARAGRAPH) 
                && (line.length() <= 2
                    || line.charAt(0) != ' ')) {
                mode = ParseMode.NORMAL; // then this line is not part of the para.
                s += "</p>" + NL; 
            } else {
                s += " "; // otherwise, need to add a space when joining up the next line.
            }

            if ((mode == ParseMode.UNORDERED_LIST) 
                && (line.length() <= 2
                    || line.charAt(0) != '-')) {
                mode = ParseMode.NORMAL; // then this line is not part of the para.
                s += "</ul>" + NL; 
            }

            if ((mode == ParseMode.ORDERED_LIST) 
                && (line.length() <= 2
                    || line.charAt(0) != '#')) {
                mode = ParseMode.NORMAL; // then this line is not part of the para.
                s += "</ol>" + NL; 
            }

            if (line.length() > 2) {
                switch (line.charAt(0)) {
                    case ' ':   // FIXME why can't i use settings.LC_PARA here?
                        if (mode != ParseMode.PARAGRAPH) {
                            s += "<p>";
                            mode = ParseMode.PARAGRAPH;
                        }
                        s += line.substring(2);
                        break;
                    case '1':   // H1
                        s += "<h1>" + line.substring(2) + "</h1>";
                        s += NL;
                        break;
                    case '2':   // H2
                        s += "<h2>" + line.substring(2) + "</h2>";
                        s += NL;
                        break;
                    case '3':   // H3
                        s += "<h3>" + line.substring(2) + "</h3>";
                        s += NL;
                        break;
                    case '4':   // H4
                        s += "<h4>" + line.substring(2) + "</h4>";
                        s += NL;
                        break;
                    case '+':   // img
                        //  TODO change the two paths to reference settings obj
                        String[] tokens = line.substring(2).split(" / ");
                        String[] filename = tokens[0].split("\\.");
                        if (tokens.length > 1) {
                            s += HTMLFigure(filename[0], filename[1], "../media/other/", tokens[1]);    
                        } else {
                            s += HTMLFigure(filename[0], filename[1], "../media/other/", "");
                        }
                        s += NL;
                        break;
                    case '-':   // unordered list
                        if (mode != ParseMode.UNORDERED_LIST) {
                            s += "<ul>" + NL;
                            mode = ParseMode.UNORDERED_LIST;
                        }
                        s += "<li>" + line.substring(2) + "</li>";
                        break;
                    case '#':   // ordered list
                        if (mode != ParseMode.ORDERED_LIST) {
                            s += "<ol>" + NL;
                            mode = ParseMode.ORDERED_LIST;
                        }
                        s += "<li>" + line.substring(2) + "</li>";
                        break;
                    default:
                        break;
                }
            }
        }
        //  all lines processed. check if we need to close off any <p>, <ul>, <ol>, tables, etc.
        switch (mode) {
            case UNORDERED_LIST:
                s += "</ul>";
                break;
            case ORDERED_LIST:
                s += "</ol>";
                break;
            case PARAGRAPH:
                s += "</p>";
                break;
            case NORMAL:
            default:
                break;
        }
        return s;
    }

    private String parsePiTags(String s, String regex, String tag) {
        String replacement = String.format("<%s>$1</%s>", tag, tag); 
        s = s.replaceAll(regex, replacement);
        return s;
    }

    private String parsePiLinks(String s) {
        int opening = s.indexOf("[[");
        int closing = s.indexOf("]]") + 2;

        while (opening > 0 && closing > 0) {
            String sub = s.substring(opening, closing);
            int split = sub.indexOf("][");
            if (split < 0) {
                //  link with no description
                //  check if it's a url or a page name.
                String stringStart = s.substring(0, opening);
                String url = sub.substring(2, sub.length() - 2);
                String stringEnd = s.substring(closing, s.length());
                s = stringStart + "<a href=\"" + url + "\">" + url + "</a>" + stringEnd;
            } else {
                //  link with description
                String stringStart = s.substring(0, opening);
                String url = sub.substring(2, split);
                String desc = sub.substring(split + 2, sub.length() - 2);
                String stringEnd = s.substring(closing, s.length());
                s = stringStart + "<a href=\"" + url + "\">" + desc + "</a>" + stringEnd;
            }
            opening = s.indexOf("[[");
            closing = s.indexOf("]]") + 2;
        }
        return s;
    }

    private String HTMLHeader() {
        String str = "<!DOCTYPE html>" + NL +
        "   <head>"  + NL +
        "       <title>%s - %s</title>" + NL +
        (this.getID().equals("index") ? "       <link rel=\"stylesheet\" href=\"css\\main.css\">" : "       <link rel=\"stylesheet\" href=\"..\\css\\main.css\">") + NL +
        "   </head>" + NL +
        "   <body>" + NL +
        "       <h1>%s</h1>" + NL;
        return String.format(str, title, settings.siteName, title);
    }

    public static String HTMLFigure(String filename, String extension, String path, String caption) {
        //  TODO add check to see if image file actually exists or id is "---"
        //  TODO add date back to caption?
        //  path = "../media/log/"
        //  extension = "jpg"
        //  filename = picid or image
        String s = "<figure><img src=\"" + path + filename + "." + extension + "\">";
        if (caption != null && !caption.equals("")) {
            s += "<figcaption>" + caption + "</figcaption>";
        }
        s += "</figure>";
        return s;
    }

    private String HTMLFooter() {
        final String NL = System.getProperty("line.separator");

        String str = NL + "      <div class=\"rule\"></div><hr>" + NL +
        "<div class=\"footer\"><p>id: " + id + "</p>" + NL +
        String.format("       <p>created: %s</p>\n", createdDate.toString()) + NL +
        String.format("       <p>updated: %s</p>\n", updatedDate.toString()) + NL +
        "      </div>" + NL +
        "   </body>" + NL +
        "</html>";
        return String.format(str, settings.siteName);
    }

    public String toHTMLString() {
        String s = "";
        s += HTMLHeader();

        // add header image, if one exists in logs
        System.out.println("\t" + logs.size() + " logs");
        for (LogEntry l : logs) {
            System.out.println("\t" + l.toString());
            if (!l.getPic().equals("---")) {
                if (l.getID().equals("index")) {
                    s += HTMLFigure(l.getPic(), "jpg", "./media/log/", l.getCaption());
                } else {
                    s += HTMLFigure(l.getPic(), "jpg", "../media/log/", l.getCaption());
                }
                l.setProcessed(true);
                break;
            }
        }

        s += parsePiContent();    
        
        //s += parseBody();

        //  add remaining images
        //  TODO change this to use an iterator https://www.java67.com/2018/12/how-to-remove-objects-or-elements-while-iterating-Arraylist-java.html
        for (LogEntry l : logs) {
            System.out.println("\t" + l.toString());
            if (!l.getPic().equals("---")  && l.getProcessed() == false) {
                s += HTMLFigure(l.getPic(), "jpg", "../media/log/", l.getCaption());
                l.setProcessed(true);
            }
        }

        // add remaining captions
        if (logs.size() > 0) {
            boolean logsToDisplay = false;
            for (LogEntry l : logs) {
                if (l.getCaption().length() > 0 && l.getProcessed() == false) {
                    if (logsToDisplay == false) {
                        logsToDisplay = true;
                        s += "<h2>Log entries</h2>" + NL;
                        s += "<ul>" + NL;
                    }
                    s += "<li>" + l.getCaption() + " &#8211; " + l.getDate()+ "</li>" + NL;
                    l.setProcessed(true);
                }
            }
            if (logsToDisplay == true) {
                s += "</ul>" + NL;
            }
        }

        s += HTMLFooter();
        return s;
    }

    public void toFile() {
        System.out.println("Generating " + this.getID());
        try {
            String path = this.getID().equals("index") ? "./" : settings.outputPath;    //  put the index in the root directory
            FileWriter fw = new FileWriter(path + "/" + getID() + ".html");
            // FIXME    handle crash if /site subdir does not exist.
            fw.write(toHTMLString());
            fw.close();
            //System.out.println("Successfully wrote to the file.");
        } catch (IOException e) {
            System.out.println("An error occurred.");
            e.printStackTrace();
        }
    }
}

File: src/page.py
import os

class Page:

    def __init__(self, src_file) -> None:
        self._src_file = src_file
        self._title = "Untitled page"
        file_and_ext = os.path.basename(src_file)
        self._filename = os.path.splitext(file_and_ext)[0]
        self._category = os.path.basename(os.path.dirname(src_file))
        self._references = [] # a list of images or other pages referenced.
        self._html = ""
        self.read_file()
        # self.join_lines()
        print(f"PAGE: {self._title} for file [{self._filename}]")
        self.parse_lines()
        # TODO self.check_references()

    def filename(self):
        return self._filename
    
    def title(self):
        return self._title

    def category(self):
        return self._category

    def html(self):
        return f"""
            {self.html_header()}
                <article>
                    <h1>{self.title()}</h1>
                    <section>
                        {self.html_body()}
                    </section>
                </article>
            {self.html_footer()}
            """
## INTERNAL ####################################################################

## html ------------------------------------------------------------------------

    def html_header(self):
        return f"""
            <!DOCTYPE html>
            <html lang="en" dir="ltr">
            <head>
                <meta charset="UTF-8">
                <link rel="stylesheet" href="/css/style.css" />
                <title>{self.title()} &mdash; Park Imminent</title>
            </head>
            <body>
            """

    def html_footer(self):
        return """
            </body>
            </html>
            """

    def html_body(self):
        return self._html

## file ------------------------------------------------------------------------

    def read_file(self):
        with open(self._src_file, 'rt') as f:
            self._lines = f.readlines()
            self._title = self._lines[0][2:-1]
            self._lines.pop(0)
        # for line in self._lines:
        #     print(len(line), '\t', line)

    def parse_lines(self):
        mode = None
        for line in self._lines:
            html, mode = self.parse_line(line, mode)
            self._html += html

    def parse_line(self, line, mode):
        line = format_line(line)
        html_string = ''
        
        if len(line) < 3:
            if mode == '.':
                mode = None
            elif mode in ['*', '-']:
                html_string += '</ul>\n'
                mode = None
            elif mode == '#':
                html_string += '</ol>\n'
                mode = None
            return html_string, mode

        rune = line[0]
        line = line[2:]

        if mode == '-' and rune not in ['-', '*']:
            html_string += '</ul>\n'
            mode = None
        if mode == '#' and rune != '#':
            html_string += '</ol>\n'
            mode = None
        if mode == '.' and rune != '.':
            html_string += '</p>\n'
            mode = None

        if mode is None:
            if rune in ['-', '*']:
                html_string += '<ul>'
                mode = '-'
            elif rune == '#':
                html_string += '<ol>'
                mode = '#'
            elif rune == '.':
                html_string += '<p>'
                mode = '.'

        if rune in ['-', '*', '#']:
            html_string += f'<li>{line}</li>\n'
        elif rune == '1':
            html_string += f'<h1>{line}</h1>\n'
        elif rune == '2':
            html_string += f'<h2>{line}</h2>\n'
        elif rune == '3':
            html_string += f'<h3>{line}</h3>\n'
        elif rune == '4':
            html_string += f'<h4>{line}</h4>\n'
        elif rune == '+':
            html_string += parse_img(line)
        elif rune == ' ':
            html_string += f'<p>{line}</p>\n'
        elif rune == '.':   # preserve linebreaks, for poetry.
            html_string += f'{line}<br>\n'
        elif rune == 'A':   # aside... should this have para formatting?
            html_string += f'<span class="right">{line}</span>\n'
        elif rune == '%':
            html_string += parse_command(line)
        else:
            html_string += f'<p style="color:green;"><b>???</b> {line}</p>\n'
        
        return html_string, mode

    def check_references():
        # check all linked files exist.
        # report errors otherwise.
        pass

## html ------------------------------------------------------------------------

def parse_img(line):
    # count / characters at the start
    num_slashes = len(line) - len(line.lstrip('/'))
    line = line.lstrip('/').lstrip()
    img_file, caption = (line.split(',', 1) + [None])[:2]
    # img_file, style, caption = (line.split(',', 2) + [None])[:3]
    img_class = {
        0: 'left',
        1: 'right',
        2: 'fullwidth'
    }
    if caption is None:
        html_string = f"""
            <figure class="{img_class[num_slashes]}">
                <img src="media\other\{img_file}">
            </figure>\n
            """
    else:
        html_string = f"""
            <figure class="{img_class[num_slashes]}">
                <img src="media\other\{img_file}">
                <figcaption>{caption}</figcaption>
            </figure>\n
            """
    return html_string

def parse_command(line):
    command_end = line.find(':')
    parsed_line = ''
    if command_end == -1:
        print('INVALID COMMAND FORMAT: ', line)
        return None
    else:
        command = line[:command_end]
        line = line[command_end + 1:].lstrip()
        match command:
            case 'html':
                parsed_line = line
            case 'ruby':
                # split on comma to get ruby pairs
                # split on colon to get ruby and ruby tag
                rubies = line.split(',')
                if len(rubies) < 1:
                    print('INVALID RUBY COMMAND: ', line)
                else:
                    for ruby in rubies:
                        if ruby.find(':') == -1:
                            print('INVALID RUBY COMMAND: ', line)
                            return None
                        else:
                            rb, rt = ruby.split(':', 1)
                            parsed_line += f'<ruby>{rb}<rp>(</rp><rt>{rt}</rt><rp>)</rp></ruby>'
            case 'link':
                comma_at = line.find(',')
                if comma_at == -1:
                    # url only
                    parsed_line = f'<a href="{line}">{line}</a>'
                else:
                    # url and description
                    parsed_line = f'<a href="{line[:comma_at]}">{line[comma_at + 1:].strip()}</a>'
            case _:
                print('UNKNOWN COMMAND: ', line)
                return None
        return parsed_line

def format_line(line):

    while has_command(line):
        com_start = line.find('{{')
        com_end = line.find('}}')
        if com_end == -1:
            print('INVALID COMMAND FORMAT: ', line)
            return line
        cmd_substring = line[com_start + 2: com_end]

        if com_end + 3 == len(line):
            line = line[:com_start] + parse_command(cmd_substring)
        else:
            line = line[:com_start] + parse_command(cmd_substring) + line[com_end + 2:]
        

    while has_link(line):
        link_start = line.find('[[')
        link_end = line.find(']]')
        if link_end == -1:
            print('INVALID LINK FORMAT: ', line)
            return line
        link_mid = line.find('][')

        if link_mid == -1 or (link_mid > link_end):
            url = line[link_start + 2: link_end]
            line = line[:link_start] + '<a href="' + url + '">' + url + '</a>' + line[link_end + 2:]
        else:
            url = line[link_start + 2: link_mid]
            label = line[link_mid + 2: link_end]
            line = line[:link_start] + '<a href="' + url + '">' + label + '</a>' + line[link_end + 2:]

    while has_italics(line):
        start = line.find('__')
        end = line.find('__', start + 2)
        text = line[start + 2: end]
        line = line[:start] + '<i>' + text + '</i>' + line[end + 2:]

    while has_bold(line):
        start = line.find('**')
        end = line.find('**', start + 2)
        text = line[start + 2: end]
        line = line[:start] + '<b>' + text + '</b>' + line[end + 2:]

    return line

def has_command(line):
    return '{{' in line

def has_link(line):
    return '[[' in line

def has_italics(line):
    return '__' in line

def has_bold(line):
    return '**' in line